reducer/src/main/java/com/graphicsfuzz/reducer/reductionopportunities/LiteralToUniformReductionOpportunity.java

```diff
@@ -16,27 +16,120 @@
 
 package com.graphicsfuzz.reducer.reductionopportunities;
 
-import com.graphicsfuzz.common.ast.expr.Expr;
+import com.graphicsfuzz.common.ast.IParentMap;
+import com.graphicsfuzz.common.ast.TranslationUnit;
+import com.graphicsfuzz.common.ast.decl.ArrayInfo;
+import com.graphicsfuzz.common.ast.decl.Declaration;
+import com.graphicsfuzz.common.ast.decl.InterfaceBlock;
+import com.graphicsfuzz.common.ast.expr.ArrayIndexExpr;
+import com.graphicsfuzz.common.ast.expr.ConstantExpr;
+import com.graphicsfuzz.common.ast.expr.IntConstantExpr;
+import com.graphicsfuzz.common.ast.expr.VariableIdentifierExpr;
+import com.graphicsfuzz.common.ast.type.ArrayType;
+import com.graphicsfuzz.common.ast.type.BasicType;
+import com.graphicsfuzz.common.ast.type.BindingLayoutQualifier;
+import com.graphicsfuzz.common.ast.type.LayoutQualifierSequence;
+import com.graphicsfuzz.common.ast.type.SetLayoutQualifier;
+import com.graphicsfuzz.common.ast.type.TypeQualifier;
 import com.graphicsfuzz.common.ast.visitors.VisitationDepth;
-import com.graphicsfuzz.common.util.PipelineInfo;
+import com.graphicsfuzz.common.transformreduce.ShaderJob;
+import com.graphicsfuzz.util.Constants;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
 
 public class LiteralToUniformReductionOpportunity
     extends AbstractReductionOpportunity {
 
-  private final Expr literalExpr;
-  private final PipelineInfo pipelineInfo;
+  private final ConstantExpr literalExpr;
+  private final ShaderJob shaderJob;
 
-  LiteralToUniformReductionOpportunity(Expr literalExpr,
-                                       PipelineInfo pipelineInfo,
+
+  LiteralToUniformReductionOpportunity(ConstantExpr literalExpr,
+                                       ShaderJob shaderJob,
                                        VisitationDepth depth) {
+
     super(depth);
+
     this.literalExpr = literalExpr;
-    this.pipelineInfo = pipelineInfo;
+    this.shaderJob = shaderJob;
+
   }
 
   @Override
   void applyReductionImpl() {
-    // TODO: ...
+
+    final String arrayName = Constants.INT_LITERAL_UNIFORM_VALUES;
+    final List<Number> values;
+    final int index;
+
+    // Creates a combined list of the current and previously added values.
+    // Updates the index number to point at the corresponding element in the array.
```

Comment:
I suggest you first check whether the array exists in the pipeline info and add an empty array if it does not.  Then your code for either finding an existing element or adding a new element will work for both cases.

Reply:
Changed accordingly.